### ISSUE TITLE
Orleans for primary/stand-by servers?

### ISSUE THREAD

**Issue by logicaloud** [dotnet/orleans#7900]
I'd like to scale an application covering anything from 2 servers to many servers and Orleans seems like a good fit for the "many servers" scenario.

**For just two servers**: Emphasis would be on redundancy, not load-balancing, and ideally all grains would be created on the primary server with the stand-by server taking over only in case of a failure of the primary server. Any shutdown or crash of the stand-by server should not affect the primary server at all, that is, there should neither be an increase in load nor any delays or temporary loss of functionality.

Can this currently be achieved with Orleans? 

With my limited knowledge, it looks like the hash-based grain placement could be used to place grains onto the primary server only, is that correct?

After a primary server failure the stand-by server should take over. This would be, as far as I know, already the standard Orleans behavior. 

Once the primary server is back up, all grains should move from the stand-by server back to the primary server. It looks like, shutting down the stand-by server with the primary server running would move grains back to the primary server although it would not be completely seamless and is not ideal. Is there a better way?

I'm aware of proposal #7692, which is great, and this would presumably allow moving grains from unwanted placement on the stand-by server back to the primary server. I haven't figured out how the stand-by server can detect that the primary server is back online within Orleans but that should be a relatively minor issue, i.e. it could be one notification away. 

If the the redundant primary/stand-by use case is not yet covered, could this be considered in the future?

**Comment by JorgeCandeias** [comment]
What is the benefit you're looking in a primary/secondary model versus the benefits Orleans already provides out-of-the-box with its distributed peer model? Is this only about not letting a "primary" server suffer noise from the collapse of a "secondary" server, which is doing no work in the first place? And wouldn't this model induce an extraordinary price when this "primary" server would collapse, requiring relocation of everything? Or is there another core reason you're looking for such a model, as opposed to just having a cluster with many nodes and letting Orleans do what it was designed to do?

**Comment by logicaloud** [comment]
> What is the benefit you're looking in a primary/secondary model versus the benefits Orleans already provides out-of-the-box with its distributed peer model? 

The benefit would be that you'd know that either server can take on the full load; there will be no surprises when one of the servers dies because you've been operating with one server anyway. You could also take down the stand-by server for maintenance without any impact (and swap roles and then maintain the primary server). 

> And wouldn't this model induce an extraordinary price when this "primary" server would collapse, requiring relocation of everything?

The number of grains that would need relocating doubles vs random placing if grains were evenly distributed between the two servers, so yes, more work in this case; not sure if that should be an issue..

> Or is there another core reason you're looking for such a model, as opposed to just having a cluster with many nodes and letting Orleans do what it was designed to do?

The primary motivation is to introduce redundancy and scalability for an application, starting with 2 servers that need to run on-premises. That means, automatically spooling up instances in the cloud is not an option. Having some virtual machine management in-house that would allow doing the same should not be mandatory.

I'm not really sure how much of a difference primary/stand-by vs load-balanced makes (being able to offer both options would be great). Maybe its more about educating users regarding load monitoring and defining some thresholds that, once breached, they know that their system may fail if one of the servers dies? Is there any monitoring in Orleans that could warn about that? 

You could argue of course that the primary server could also become overloaded, but usually issues build up over time with warning signs and time to act vs a single, sudden incident.

It may be my lack of knowledge that makes me see a primary/stand-by setup as the safer option for a two server system.

**Comment by benjaminpetit** [comment]
The default grain directory is distributed. So, if one silo dies unexpectedly in a cluster, you might see transient errors even if this silo had no grain activations.

> It looks like, shutting down the stand-by server with the primary server running would move grains back to the primary server although it would not be completely seamless and is not ideal. Is there a better way?

It should be seamless if the shutdown is graceful.
Otherwise, you could just let them migrate over time when activations become idle.

I am not convinced that having standby silos is a good design. Better to use all resource you have, you will have less disruption in case of a silo crash. If you really want to handle all your load on one server, then you will have to dimension it big enough.

> Maybe its more about educating users regarding load monitoring and defining some thresholds that, once breached, they know that their system may fail if one of the servers dies? Is there any monitoring in Orleans that could warn about that?

I guess it would depend on your app? If you are CPU bound, then set a threshold on that. If you have a reliable perf test scenario, you could get a safe maximum number of activations, etc.

**Comment by JorgeCandeias** [comment]
Thanks for sharing the motivation behind this. 

TLDR; There's no need to add complexity with an explicit primary/secondary model to achieve full redundancy, even with only two servers. Orleans will manage redundancy for us, all we need to do is to manage capacity.

Long story:

To achieve this, we can just use the hosting stack to monitor resource usage. If we want to ensure there is always capacity on a two node setup, then we ensure that average resource usage in the cluster never meets or exceeds 1/2 (50%). On a three node setup, even if two nodes die, then ensure <1/3 (33%). And so on. Generally such absolute redundancy is not required with a large cluster. We can keep average resource usage under a reasonable value, and the cluster will tolerate a few nodes dying at the same time, long enough for other nodes to spin up - either automatically via the hosting stack (e.g. kubernetes), or manually via restarting a faulty box or service (e.g. windows service).

Note 1: The most critical resource in Orleans is RAM. Orleans can run out of CPU capacity and still do a very fair job of managing load while we address that problem. Yet running out of cluster memory while the system is under heavy load, and while also not being able to spin up new nodes to replace dead ones, can trigger a snowball effect that can take the entire cluster down. This is true for other in-memory actor systems it is a conscious trade-off when using such technology. It must be planned for, which is what you are correctly attempting to do here.

Note 2: Having only two servers in the cluster can invite problems during recovery. The reason is the voting mechanism. By default, the cluster is looking for at least two votes that a silo is dead before kicking it out. The faster you get other silos agreeing that a particular silo is dead, the faster that silo gets kicked out. So the more nodes you have in the cluster, the faster recovery is. With a two node setup where one node dies, there is only one voting node left there. The algorithm is smart enough to allow a single node to eventually kick all other nodes out, but for safety reasons, it can taker a very long time. This vote count can be configured if necessary.

Note 3: Having an even number of servers generally invites trouble for any distributed cluster. The reason is that there is no natural fallback to solve a split brain condition with a pair number of servers, if the split happens right down the middle. Since in a physical setup, we may well have half of those servers on a rack, half on the other, this condition is not as rare as it sounds. Each side of the split will mark the other side as dead and there won't be a clear winner. Whereas if you have an odd number of servers, the system can always fallback on whatever side is largest.

Note 4: You will still see errors even if the "secondary" node dies, because they are still forming a cluster.

So my suggestion is to not look at implementing a primary/secondary model at all. It will not only require a lot of custom code for placement directors, it is also unnecessary and inefficient, given what Orleans offers out of the box. It's easier and more efficient to just use all servers and ensure the average capacity (especially RAM) is under an appropriate threshold for a given failure tolerance level. If that threshold is exceeded, look at adding capacity to bring the average below the threshold again. Also, rule of thumb, the more nodes we have, the higher the threshold can be, as when one node fails, its load will be spread out across the cluster. But we must beware the snowball effect if our setup does not allow additional nodes to be provisioned quickly.

**Comment by logicaloud** [comment]
@benjaminpetit @JorgeCandeias Thank you for your comments and the detailed explanation. My conclusions are

1. Orleans is not a good fit for a two-server setup; an active/passive two-server setup is not supported. 
2. Development of features (either outside or within the Orleans sources) covering the specific active/passive two-server case is not worth the effort.

Looks like the application should go from one to three servers to start with; I'll be experimenting with that a little.

I'll leave the issue open for a day or two in case someone else has any comments but happy to close otherwise.

**Comment by benjaminpetit** [comment]
There is no issue running a two-server cluster with Orleans.

Are you afraid of a split-brain scenario? It should not happen when using a production-grade membership provider (like Azure Table, or SQL)

**Comment by logicaloud** [comment]
@benjaminpetit It sounded like that the split-brain scenario could cause friction, i.e. delays in recovery? Is there any quantitative measure for that? Then I could judge if it is acceptable. I'll definitely try myself once I get that far.

**Comment by benjaminpetit** [comment]
For Orleans membership, the source of truth is in the membership table.

Let's say you have a cluster with `SiloA` and `SiloB`. `SiloB` crashes unexpectedly. `SiloA` will see that `SiloB` isn't responding anymore, and will start a vote to declare `SiloB` ad `Dead`.

Since there are only two silos in your cluster, one vote is sufficient to declare a silo `Dead`. So `SiloA` will then update the membership table, putting the entry `SiloB` as `Dead`. Operations will continue as normal.

Now let's say you had a temporary network issue, that cut communications between your two silos (but they are still running fine). They might try to vote each other `Dead`, but only one will win and be able to update the membership table. The silo that was too slow to update the memebrship table will see its entry marked as `Dead` and will initiate a fast shutdown.

**Comment by logicaloud** [comment]
Thank you; I'll close and start experimenting.
